### DISCUSSION TITLE
Overriding PersistentMarker's Y position

### DISCUSSION THREAD

**Discussion by Benyuss** [patrykandpatrick/vico#622] [Questions]
Hello @patrickmichalik !
I once again looking for your help.
Maybe it's just me but I sometimes have serious issues understanding the library and can't do simple tasks like the following by myself.. (even with a decade of experience..) Is there a way to document your library better? Are you looking for any help?

![Screenshot 2024-03-19 at 19 53 21](https://github.com/patrykandpatrick/vico/assets/18491308/c118afa3-787e-4b55-9375-a05a7bdbfa54)
I'd like to create a persistent marker on the first position in the chart. The problem is, the chart doesn't use this position at all (I'm not using insets or anything just a regular persistentmarker, key is X position of the first entry.
The marker always lands in the same exact position, no matter where the chart starts.
Is there a way to achive this behaviour?

Here's a pic about the current status:
![Screenshot 2024-03-19 at 19 57 50](https://github.com/patrykandpatrick/vico/assets/18491308/141223c9-9826-4ce1-a1af-91da99d471de)
(also sometimes X position labels get cut-off but that's another issue that I'll bring up in a separate topic((if you want to)))

Thanks in advance!

**Comment by Gowsky** [comment]
Hi! First of all, please provide a complete description of your problem:
- What Vico version are you using?
- Are you using Compose or views?
- How is the `Marker` created? Are you using `MarkerComponent`?

>Is there a way to document your library better? Are you looking for any help?

Presumably yes, but we have not received any suggestions yet. Specific suggestions would be more helpful than vague complaints.

>(even with a decade of experience..)

I don't think that's relevant. But I'm sure that with so much experience you know that you can read the source code and make overrides, and you know what information should be provided when asking for help.

>(also sometimes X position labels get cut-off but that's another issue that I'll bring up in a separate topic((if you want to)))

Sure, please do.

**Comment by patrickmichalik** [comment]
Hello, @Benyuss! I presume you’re using `MarkerComponent`—please let me know if this isn’t the case. In [Vico 2.0.0 Alpha 7](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.7), we introduced label position customization for `MarkerComponent`. `MarkerComponent.LabelPosition.AbovePoint` positions the label above the marked point rather than at the top of the `CartesianChart`, the latter behavior being produced by `MarkerComponent.LabelPosition.Top`. There’s also `MarkerComponent.LabelPosition.AroundPoint`. When this is applied, no insets are requested, and the label is positioned below the marked point when there isn’t enough room above it.

Below are two examples with `MarkerComponent.LabelPosition.Top`, one for Jetpack Compose and one for the view system. The negative bottom margin shifts the label down by half of the line width of the `LineCartesianLayer` to which the `MarkerComponent` will be added. This prevents a sliver of the line from sticking out.

```kt
rememberMarkerComponent(
    rememberTextComponent(
        color = Color.White,
        background = rememberShapeComponent(Shapes.roundedCornerShape(4.dp), Color.Gray),
        padding = dimensionsOf(8.dp, 4.dp),
        margins = dimensionsOf(bottom = (-1).dp),
    ),
    remember { DefaultMarkerLabelFormatter(colorCode = false) },
    MarkerComponent.LabelPosition.AbovePoint,
)
```

```kt
MarkerComponent(
    TextComponent.build {
        color = Color.WHITE
        background = ShapeComponent(Shapes.roundedCornerShape(4f), Color.GRAY)
        padding = dimensionsOf(horizontalDp = 8f, verticalDp = 4f)
        margins = dimensionsOf(bottomDp = -1f)
    },
    DefaultMarkerLabelFormatter(colorCode = false),
    MarkerComponent.LabelPosition.AbovePoint,
)
```

Adding this to the first sample chart as a persistent `Marker`, we get the following:

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/b648a67f-53d8-4b03-9e62-214cd3ff47b0" width="720" />

Such a `Marker` can be created in Vico 1 too, but a custom `Marker` implementation is required, as `MarkerComponent.LabelPosition` isn’t available. Let me know if you need guidance for Vico 1 or if you have any other questions.

It’s generally good to keep in mind that `Marker` is an interface—where `MarkerComponent` is insufficient, you can create a custom implementation to draw anything at any position. It should also be noted that in Vico 2, `MarkerComponent` subclasses can be created, which is useful for small overrides.

We’re sorry to hear that you find the documentation insufficient. We’ll look into making the descriptions of `Marker` and `MarkerComponent` more comprehensive. Your input on what other areas should be better documented would be very helpful. Thanks!
